### ISSUE TITLE
amp-story can only play 8 audio or 16 video elements

### ISSUE THREAD

**Issue by smndhm** [ampproject/amphtml#13476]
## What's the issue?

background-audio in amp-story only works for the 8 firsts items.

## How do we reproduce the issue?

Create a 9+ pages amp-story.

## What browsers are affected?

All browsers, desktop and mobile.

## Which AMP version is affected?

Version 1518441587106

**Comment by newmuis** [comment]
Thanks for the report @smndhm!  Was this using the `background-audio` attribute on an `amp-story` element that has 9+ `amp-story-page` children?  Or did you put the `background-audio` attribute directly on 9+ `amp-story-page` elements?

**Comment by smndhm** [comment]
I'm using it on the `amp-story-page`

**Comment by newmuis** [comment]
👍  Thanks @smndhm.  I'll investigate.  My first guess is to blame the [`MediaPool`](https://github.com/ampproject/amphtml/blob/master/extensions/amp-story/0.1/media-pool.js).

**Comment by newmuis** [comment]
/cc @Refurb who had this issue in #14624

@alanorozco does the video refactor affect audio at all?  I'm wondering whether we are good to start investigating this now, or whether it makes sense to wait until after the video refactor lands.

**Comment by ampprojectbot** [comment]
This is a high priority issue but it hasn't been updated in awhile. @newmuis Do you have any updates?

**Comment by alanorozco** [comment]
@newmuis

Sorry for missing this, got lost in the shuffle.

I don't see how the video refactor interacts with this issue, could you explain further?

**Comment by newmuis** [comment]
After understanding the video refactor better, I don't think it affects this either.

**Comment by ampprojectbot** [comment]
This is a high priority issue but it hasn't been updated in awhile. @newmuis Do you have any updates?

**Comment by ampprojectbot** [comment]
This is a high priority issue but it hasn't been updated in awhile. @newmuis Do you have any updates?

**Comment by ampprojectbot** [comment]
This is a high priority issue but it hasn't been updated in awhile. @newmuis Do you have any updates?

**Comment by ampprojectbot** [comment]
This is a high priority issue but it hasn't been updated in awhile. @newmuis Do you have any updates?

**Comment by ampprojectbot** [comment]
This is a high priority issue but it hasn't been updated in awhile. @newmuis Do you have any updates?

**Comment by tomxbarnesx** [comment]
I'm having similar issues to the original poster. I was only attempting to upload 5 background audio files. The files are arranged correctly in the amp html and are accessible in the src property via the developer tools, but nothing plays.

I noticed it while trying to build this new story, but going back through our archives, it seems that all the slides that have background audio (all as attributes within an amp-story-page, all working up until very recently) are no longer functional. 

Has something changed with the way amp is serving/playing background audio from story-page attributes?

**Comment by newmuis** [comment]
Nothing has changed, but the state is also dependent on what else is happening on your machine, especially if you are using an Apple device.  If you have other tabs or applications open that have audio or video (whether they are playing or not), it will decrease the number of audio or video elements you can use in your story.

Of course, the goal is to mitigate this as much as possible, and that is what this issue tracks.  Ideally, we would be able to play your story's media as long as there is any slot available on your machine to do so, and ideally we would provide a graceful fallback in the case where you have no hardware decoding slots left available to play the media.

This is something that we are actively working on this quarter.

**Comment by newmuis** [comment]
For the record, we have since discovered that the issue that @tomxbarnesx describes here was actually a separate issue.  #20473 has context.

**Comment by tomxbarnesx** [comment]
Thanks so much for the updates. Yeah I had a feeling the problem wasn't entirely related to hardware slotting. Good luck with the future updates

**Comment by ampprojectbot** [comment]
This is a high priority issue but it hasn't been updated in awhile. @newmuis Do you have any updates?

**Comment by newmuis** [comment]
This affects both audio and video.  If `n` is the size of the MediaPool for the specified media type, then the you will encounter an error once you navigate to the page containing the `2n+1`th media element of that type.

This is because we delay clear the associated `replaced-media` after the promise from the `SwapOutOfDomTask` resolves:

https://github.com/ampproject/amphtml/blob/969a0a0ea9d093010f03980cbf61a898970b248a/extensions/amp-story/1.0/media-pool.js#L570-L574

The `SwapOutOfDomTask` is executed in a microtask.  However, when this media element is being swapped back into the DOM, there is synchronous code that sets the `replaced-media` before the `SwapIntoDomTask` resolves:

https://github.com/ampproject/amphtml/blob/969a0a0ea9d093010f03980cbf61a898970b248a/extensions/amp-story/1.0/media-pool.js#L502-L514

As such, we get this order of events:

1. `i-amphtml-pool-media-4` is replacing `i-amphtml-placeholder-media-0`
2. setting `replaced-media` of `i-amphtml-pool-media-4` to `i-amphtml-placeholder-media-0`
3. swapping `i-amphtml-pool-media-4` into DOM complete
4. `i-amphtml-placeholder-media-0` is being put back in DOM; `i-amphtml-pool-media-4` is now available for use
5. `i-amphtml-pool-media-4` is replacing `i-amphtml-placeholder-media-3`
6. setting `replaced-media` of `i-amphtml-pool-media-4` to `i-amphtml-placeholder-media-3`
7. swapping `i-amphtml-pool-media-4` out of DOM complete
8. clearing `replaced-media` of `i-amphtml-pool-media-4`
9. swapping `i-amphtml-pool-media-4` into DOM complete
10. ERROR: `i-amphtml-pool-media-4` has no associated placeholder

As you can see from the logs above, this is problematic because even though the element is swapped out of the DOM (step 4) before it is swapped back in (step 5), the old `replaced-media` value is cleared (step 8) *after* it is set to the new value (step 6).

The simplest fix here is to set the `replaced-media` value after the promise for the  `SwapIntoDomTask` completes (i.e. move step 6 after step 9).

I will send a PR to fix this.
